### ISSUE TITLE
if-let borrow then clone unexpected compiler error

### ISSUE THREAD

**Issue by l4l** [rust-lang/rust#77623]
With that definition:

```rust
#[derive(Default)]
struct T {
    x: Option<Rc<RefCell<T>>>,
}
```

The following code fails to compile:

```rust
let mut t = Rc::new(RefCell::new(T { x: Some(Rc::default()) }));
if let Some(v) = t.borrow().x.clone() {
    t = v;
};

// compile error:
error[E0506]: cannot assign to `t` because it is borrowed
  --> src/main.rs:19:9
   |
18 |     if let Some(v) = t.borrow().x.clone() {
   |                      ----------
   |                      |
   |                      borrow of `t` occurs here
   |                      a temporary with access to the borrow is created here ...
19 |         t = v;
   |         ^ assignment to borrowed `t` occurs here
20 |     };
   |      - ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `std::cell::Ref<'_, main::T>`
```

Though if we introduce an additional value binding this magically works:

```rust
let mut t = Rc::new(RefCell::new(T { x: Some(Rc::default()) }));
if let Some(v) = { let x = t.borrow().x.clone(); x } {
    t = v;
};
```

### Meta

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7d3700c4dc0f27a65cf18025f2307e34) (rustc 1.46 on stable branch)

**Comment by jonas-schievink** [comment]
This is working as intended. The temporary `Ref` is dropped at the end of the containing statement, which is the `if let`.

**Comment by jonas-schievink** [comment]
(this is documented here: https://doc.rust-lang.org/reference/destructors.html#temporary-scopes)
